### PR TITLE
feat: Argocd App Unset Kustomize Override

### PR DESCRIPTION
**Pull request by mayzhang2000**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

File: cmd/argocd/commands/app.go [hunk 1]
```diff
@@ -777,7 +776,7 @@ func NewApplicationUnsetCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
 			updated := false
 			if app.Spec.Source.Kustomize != nil {
 				for _, paramStr := range parameters {
-					unsetKustomizeOpt(&app.Spec.Source, paramStr);
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]: I think `argocd app unset` flags we should use consistent with `argocd app set` flags. 
Can you please introduce `--kustomize-image`, `--nameprefix` and `namesuffix` . Flags are defined here: https://github.com/argoproj/argo-cd/blob/ebc048167c2e86f3e119eadb69fd5150ac1f7823/cmd/argocd/commands/app.go#L721

And use `flags.Visit` to iterate through specified list of flags and "unset" corresponding fields. Example it here:

https://github.com/argoproj/argo-cd/blob/ebc048167c2e86f3e119eadb69fd5150ac1f7823/cmd/argocd/commands/app.go#L473-L475

Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: Great input. Thank you Alex. Changes are checked-in.

File: cmd/argocd/commands/app.go [hunk 2]
```diff
@@ -750,7 +781,37 @@ func NewApplicationUnsetCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
 			errors.CheckError(err)
 
 			updated := false
+			if app.Spec.Source.Kustomize != nil {
+				visited := 0
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]: The `visited` variable is used in `setAppSpecOptions` to find out if anything has changed in the app. 

In the `NewApplicationUnsetCommand` we have `updated` flag. Please use it instead. E.g.:

```
c.Flags().Visit(func(f *pflag.Flag) {
    switch f.Name {
    case "nameprefix":
        updated = true
        app.Spec.Source.Kustomize.NamePrefix = ""
    case "namesuffix":
        updated = true
        app.Spec.Source.Kustomize.NameSuffix = ""
    case "kustomize-image":
        for _, kustomizeImage := range kustomizeImages {
            for i, item := range app.Spec.Source.Kustomize.Images {
                if item.Match(argoappv1.KustomizeImage(kustomizeImage)) {
                    updated = true
                    
```

File: cmd/argocd/commands/app.go [hunk 3]
```diff
@@ -729,17 +729,48 @@ func addAppFlags(command *cobra.Command, opts *appOptions) {
 	command.Flags().StringArrayVar(&opts.kustomizeImages, "kustomize-image", []string{}, "Kustomize images (e.g. --kustomize-image node:8.15.0 --kustomize-image mysql=mariadb,alpine@sha256:24a0c4b4a4c0eb97a1aabb8e29f18e917d05abfe1b7a7c07857230879ce7d3d)")
 }
 
+//Usage:
+//app unset <appName> -p kustomize-image=alpine
+//app unset <appName> -p namesuffix
+//app unset <appName> -p nameprefix
+func unsetKustomizeOpt(src *argoappv1.ApplicationSource, paramStr string) {
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]: Looks like this function is not used anymore. Please remove it

File: cmd/argocd/commands/app.go [hunk 4]
```diff
@@ -732,14 +732,17 @@ func addAppFlags(command *cobra.Command, opts *appOptions) {
...
-		parameters  []string
-		valuesFiles []string
+		parameters      []string
+		valuesFiles     []string
+		nameSuffix      string
+		namePrefix      string
+		kustomizeImages []string
 	)
 	var command = &cobra.Command{
-		Use:   "unset APPNAME -p COMPONENT=PARAM",
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]: Can you please add usage example. Similar to https://github.com/argoproj/argo-cd/blob/697e692c0caa832f192ff6734043d8c1f9c6dfdd/cmd/argocd/commands/app.go#L1373-L1377

File: cmd/argocd/commands/app.go [hunk 5]
```diff
@@ -732,14 +732,17 @@ func addAppFlags(command *cobra.Command, opts *appOptions) {
...
-		Short: "Unset application parameters",
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]: Just "Unset application" is not very clear . I think `Unset application parameters` is good enough. We keep referring to config tool settings as "parameters".

**Review by alexmt — CHANGES_REQUESTED** [review]

File: cmd/argocd/commands/app.go (hunk 3 above)
Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: This is already removed in https://github.com/argoproj/argo-cd/pull/3289/commits/0fbcabd54da28b862c7a2e048bf4d2c354996952

File: cmd/argocd/commands/app.go (hunk 5 above)
Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: :+1 Change is checked in.

File: cmd/argocd/commands/app.go (hunk 4 above)
Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: :+1: Code is checked in

File: cmd/argocd/commands/app.go (hunk 2 above)
Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: Fixed.

File: cmd/argocd/commands/app.go [hunk 6]
```diff
@@ -750,7 +762,38 @@ func NewApplicationUnsetCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+					case "nameprefix":
+						updated = true
+						app.Spec.Source.Kustomize.NamePrefix = ""
+					case "namesuffix":
+						updated = true
+						app.Spec.Source.Kustomize.NameSuffix = ""
+					case "kustomize-image":
+						for _, kustomizeImage := range kustomizeImages {
+							for i, item := range app.Spec.Source.Kustomize.Images {
+								if item.Match(argoappv1.KustomizeImage(kustomizeImage)) {
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]: Noticed one issue. This check should be implemented as follows (kustomize set expression checks for image name, not visa versa):

```suggestion
								if argoappv1.KustomizeImage(kustomizeImage).Match(item) {
```

File: cmd/argocd/commands/app.go [hunk 7]
```diff
@@ -802,6 +849,9 @@ func NewApplicationUnsetCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
 	}
 	command.Flags().StringArrayVarP(&parameters, "parameter", "p", []string{}, "unset a parameter override (e.g. -p guestbook=image)")
 	command.Flags().StringArrayVar(&valuesFiles, "values", []string{}, "unset one or more helm values files")
+	command.Flags().StringVar(&nameSuffix, "namesuffix", "", "Kustomize namesuffix")
+	command.Flags().StringVar(&namePrefix, "nameprefix", "", "Kustomize nameprefix")
+	command.Flags().StringArrayVar(&kustomizeImages, "kustomize-image", []string{}, "Kustomize images (e.g. --kustomize-image node:8.15.0 --kustomize-image mysql=mariadb,alpine@sha256:24a0c4b4a4c0eb97a1aabb8e29f18e917d05abfe1b7a7c07857230879ce7d3d)")
```
Comment (alexmt) [inline cmd/argocd/commands/app.go]:
```suggestion
	command.Flags().StringArrayVar(&kustomizeImages, "kustomize-image", []string{}, "Kustomize images name (e.g. --kustomize-image node --kustomize-image mysql)")
```

File: test/e2e/kustomize_test.go [hunk 8]
```diff
@@ -194,3 +214,22 @@ func TestKustomizeNameSuffix(t *testing.T) {
 			assert.Contains(t, app.Spec.Source.Kustomize.NameSuffix, "-suf")
 		})
 }
+
+// make sure we we can invoke the CLI to set and unset namesuffix
+func TestKustomizeNameSuffixSetUnset(t *testing.T) {
```
Comment (alexmt) [inline test/e2e/kustomize_test.go]: Can you merge TestKustomizeImagesSetUnset and TestKustomizeNameSuffixSetUnset into one test. The e2e tests are quite expensive. I think it this case it is better to have one test that verifies both parameters

**Review by alexmt — CHANGES_REQUESTED** [review]

File: cmd/argocd/commands/app.go (hunk 6 above)
Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: Fixed.

File: cmd/argocd/commands/app.go (hunk 7 above)
Comment (mayzhang2000) [reply cmd/argocd/commands/app.go]: Checked in

File: test/e2e/kustomize_test.go (hunk 8 above)
Comment (mayzhang2000) [reply test/e2e/kustomize_test.go]: Two tests are merged into one. Thank you Alex for the detailed and thorough review!

**Review by alexmt — APPROVED** [review]
LGTM!

**Pull request merged**